src/DIRAC/WorkloadManagementSystem/DB/VirtualMachineDB.py

```diff
@@ -0,0 +1,1429 @@
+      return result
+    if not result['Value']:
+      return S_ERROR('Instance not found')
+    uniqueID = result['Value'][0][0]
+
+    return S_OK(uniqueID)
+
+  def getInstanceID(self, uniqueID):
+    """
+    For a given uniqueID of an instance return associated internal InstanceID
+    """
+    tableName, _validStates, idName = self.__getTypeTuple('Instance')
+
+    result = self.getFields(tableName, [idName], {'UniqueID': uniqueID})
+    if not result['OK']:
+      return result
+    if not result['Value']:
+      return S_ERROR('Unknown %s = %s' % ('UniqueID', uniqueID))
+    return S_OK(result['Value'][0][0])
+
+  def declareInstanceSubmitted(self, uniqueID):
+    """
+    After submission of the instance the Director should declare the submitted Status
+    """
+    instanceID = self.getInstanceID(uniqueID)
+    if not instanceID['OK']:
+      return instanceID
+    instanceID = instanceID['Value']
+
+    status = self.__setState('Instance', instanceID, 'Submitted')
+    if status['OK']:
+      self.__addInstanceHistory(instanceID, 'Submitted')
+
+    return status
+
+  def setPublicIP(self, instanceID, publicIP):
+    """
+    Update publicIP when used for contextualization previus to declareInstanceRunning
+    """
+    publicIP = publicIP.replace('::ffff:', '')
+    result = self._escapeString(publicIP)
+    if not result['OK']:
+      return result
+    publicIP = result['Value']
+
+    try:
+      instanceID = int(instanceID)
+    except ValueError:
+      return S_ERROR("instanceID has to be an integer value")
+
+    tableName, _validStates, idName = self.__getTypeTuple('Instance')
+    sqlUpdate = 'UPDATE `%s` SET PublicIP = %s WHERE %s = %d' % (tableName, publicIP, idName, instanceID)
+
+    return self._update(sqlUpdate)
+
+  def declareInstanceRunning(self, uniqueID, publicIP, privateIP=""):
+    """
+    Declares an instance Running and sets its associated info (uniqueID, publicIP, privateIP)
+    Returns S_ERROR if:
+    - instanceName does not have a "Submitted" or "Contextualizing" entry
+    - uniqueID is not unique
+    """
+    instanceID = self.getInstanceID(uniqueID)
+    if not instanceID['OK']:
+      return instanceID
+    instanceID = instanceID['Value']
+
+    # No IPv6 prefix
+    publicIP = publicIP.replace('::ffff:', '')
+
+    self.__setInstanceIPs(instanceID, publicIP, privateIP)
+
+    status = self.__setState('Instance', instanceID, 'Running')
+    if status['OK']:
+      self.__addInstanceHistory(instanceID, 'Running')
+
+    return self.getAllInfoForUniqueID(uniqueID)
+
+  def declareInstanceStopping(self, instanceID):
+    """
+    Mark a VM instance to be stopped.
+
+    Next time the instance's VirtualMachineMonitor checks in for an update
+    it will be told to halt.
+
+    :return: S_OK if instance updated, S_ERROR otherwise.
+    """
+    status = self.__setState('Instance', instanceID, 'Stopping')
+    if status['OK']:
+      self.__addInstanceHistory(instanceID, 'Stopping')
```

Comment:
```suggestion
      return self.__addInstanceHistory(instanceID, 'Stopping')
```
?

Reply:
I think this is deliberate... The setState call si the important, so that's the return value that's used. I've added a warning log message if the addInstanceHistory call fails.